Deduplicate runtime metrics (fix #157)

knyar opened the pull request:
I've chosen to call these "runtime metrics" based on [the piece that generates them](https://github.com/cloudfoundry/dropsonde/blob/master/runtime_stats/runtime_stats.go).
Please let me know if you can come up with a better term :)

The regular expression is part of the config to avoid hard-coding it as a string elsewhere. I don't really expect anyone to override this, which is why it's not wired through to the manifest or tile.

I considered a comma-separated list of metric names instead of a regular expression, but it will be pretty long. Happy to do that if you think the performance difference matters.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/161)
<!-- Reviewable:end -->

johnsonj commented:
I think the regex approach is great and have no concerns performance wise. We do more expensive stuff in the hot path (constantly allocating, filling map[string]string) and it hasn't been identified as an issue. Agree with the config option (as mentioned below) 

Only feedback is code structure, otherwise it looks great.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/stackdriver-nozzle/config/config.go, line 69 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IxZmzhArLz6Y829lA:-L-IxZmzhArLz6Y829lB:b-mvtt0w) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/config/config.go#L69)):*
> ```Go
> 	MetricsBatchSize      int    `envconfig:"metrics_batch_size" default:"200"`
> 	MetricPathPrefix      string `envconfig:"metric_path_prefix" default:"firehose"`
> 	RuntimeMetricRe       string `envconfig:"runtime_metric_re" default:"^(numCPUS|numGoRoutines|memoryStats\\..*)$"`
> ```

I don't think people are going to need to change this (so no need for tile/spec changes until it's raised as an issue), but I like that you provide it as a configuration option because it is config. 

Can you expand the name here, eg `RuntimeMetricRegex`. Re may be common but Regex is undeniably a regular expression. 

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 35 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IxFXARm1IfUS5r6C4:-L-IxFXARm1IfUS5r6C5:b-ra3160) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L35)):*
> ```Go
> 	r, err := regexp.Compile(runtimeMetricRe)
> 	if err != nil {
> 		return nil, err
> ```

suggest wrapping in more descriptive error since a regex failure may seem a little bit random here, eg:
```go
return nil, fmt.Errorf("compiling runtime metrix regex: %v", err)
```

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 53 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-Iy1ublNwoMP4yAv3Q:-L-Iy1ublNwoMP4yAv3R:bxqo0f5) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L53)):*
> ```Go
> 	unitParser    UnitParser
> 	logger        lager.Logger
> 	// runtimeMetricRe is a regexp that a ValueMetric name is matched against.
> ```

nit/documentation: Instead of explaining the regex, put a comment above isRuntimeMetric and explain:

- why we need to identify RuntimeMetrics
- what we handle specially for RuntimeMetrics vs normal metrics
- how we're matching it (the regex) (maybe, if it feels needed) 

This comment may be better suited in the config struct

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 77 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IycfVHMH3HHMhqTUi:-L-IycfVHMH3HHMhqTUj:bfnb5lo) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L77)):*
> ```Go
> 
> func (ms *metricSink) Receive(envelope *events.Envelope) {
> 	labels := ms.labelMaker.MetricLabels(envelope)
> ```

is it difficult/too complex to do this in the `labelMaker`? this feels like it's responsibility.

if you're having trouble reconciling the separation maybe metric naming should be the responsibility of the `labelMaker` as well. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-:-L-IyytL8x7p4zirPJK9:bm9vsh2)*
<!-- Sent from Reviewable.io -->

knyar commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/stackdriver-nozzle/config/config.go, line 69 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IxZmzhArLz6Y829lA:-L-WBugP6qF3n1aBx6Pr:b-896fix) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/config/config.go#L69)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

I don't think people are going to need to change this (so no need for tile/spec changes until it's raised as an issue), but I like that you provide it as a configuration option because it is config. 

Can you expand the name here, eg `RuntimeMetricRegex`. Re may be common but Regex is undeniably a regular expression. 
</blockquote></details>

Done.

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 35 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IxFXARm1IfUS5r6C4:-L-WDQ_FJA3ios_061Yw:b-896fix) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L35)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

suggest wrapping in more descriptive error since a regex failure may seem a little bit random here, eg:
```go
return nil, fmt.Errorf("compiling runtime metrix regex: %v", err)
```
</blockquote></details>

Done.

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 53 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-Iy1ublNwoMP4yAv3Q:-L-WFaGMLiO4MNIQmQPG:bvgx7u2) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L53)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

nit/documentation: Instead of explaining the regex, put a comment above isRuntimeMetric and explain:

- why we need to identify RuntimeMetrics
- what we handle specially for RuntimeMetrics vs normal metrics
- how we're matching it (the regex) (maybe, if it feels needed) 

This comment may be better suited in the config struct
</blockquote></details>

Added a longer comment for isRuntimeMetric, and a short note in the config struct.

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 77 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IycfVHMH3HHMhqTUi:-L-WN5oq9nFbAyfSA3wK:bnoftdu) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L77)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

is it difficult/too complex to do this in the `labelMaker`? this feels like it's responsibility.

if you're having trouble reconciling the separation maybe metric naming should be the responsibility of the `labelMaker` as well. 
</blockquote></details>

`LabelMaker` also handles log labels, so I don't think it's the right place to generate metric names. However, I've added a boolean argument to `MetricLabels` which controls whether `origin` should be added, making this more explicit. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I've got a suggestion for the API. Feel free to address and submit or just submit as is. 

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 81 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXmJJ_gr6Rx5gTLZZ:-L-XXmJJ_gr6Rx5gTLZ_:b-d3vcjc) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L81)):*
> ```Go
> // We maintain vm and application instance indexes as separate labels so that
> // it is easy to aggregate across multiple instances.
> func (lm *labelMaker) MetricLabels(envelope *events.Envelope, addOrigin bool) map[string]string {
> ```

If we end up adding more options here, we should make addOrigin read more explicitly, eg:

```go
// Callers
lm.MetricLabels(envelope, nozzle.WithOrigin)
lm.MetricLabels(envelope, nozzle.WithoutOrigin)

// implementation
type OriginOption bool
var WithOrigin = OriginOption(true)
var WithoutOrigin = OriginOption(false)
```

I'll leave it up to you if you think that's appropriate for this change.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 90 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXMdtEp2o9KdbbUgC:-L-XXMdtEp2o9KdbbUgD:bjcyelk) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L90)):*
> ```Go
> 	labels.setIfNotEmpty("instanceIndex", getInstanceIndex(envelope))
> 	labels.setIfNotEmpty("tags", getTags(envelope))
> 	if addOrigin {
> ```

thumbs up!

---

*[src/stackdriver-nozzle/nozzle/metric_sink.go, line 77 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-IycfVHMH3HHMhqTUi:-L-XXbdjsmEPhjyauTTY:b9j9n3q) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/34a2f4cddd4e932a426be6d24bfa9b7d78b98e69/src/stackdriver-nozzle/nozzle/metric_sink.go#L77)):*
<details><summary><i>Previously, knyar (Anton Tolchanov) wrote…</i></summary><blockquote>

`LabelMaker` also handles log labels, so I don't think it's the right place to generate metric names. However, I've added a boolean argument to `MetricLabels` which controls whether `origin` should be added, making this more explicit. What do you think?
</blockquote></details>

Looks good to me!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-:-L-XYcfE0Z-oqe1sWRV6:b-ogrncu)*
<!-- Sent from Reviewable.io -->

knyar commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 81 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXmJJ_gr6Rx5gTLZZ:-L-XbP_dV73h11nG_N7I:b-e7xnvs) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L81)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

If we end up adding more options here, we should make addOrigin read more explicitly, eg:

```go
// Callers
lm.MetricLabels(envelope, nozzle.WithOrigin)
lm.MetricLabels(envelope, nozzle.WithoutOrigin)

// implementation
type OriginOption bool
var WithOrigin = OriginOption(true)
var WithoutOrigin = OriginOption(false)
```

I'll leave it up to you if you think that's appropriate for this change.

</blockquote></details>

Oh, this looks neat. However, it will make calling MetricLabels from metric_sink slightly more cumbersome (will need to cast `bool` into `OriginOption`), and since we only currently have a single option, I think leaving it as a simple `bool` for now is simpler.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 81 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXmJJ_gr6Rx5gTLZZ:-L-Xdbj82heX8F415zsX:b-2y0xu5) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L81)):*
<details><summary><i>Previously, knyar (Anton Tolchanov) wrote…</i></summary><blockquote>

Oh, this looks neat. However, it will make calling MetricLabels from metric_sink slightly more cumbersome (will need to cast `bool` into `OriginOption`), and since we only currently have a single option, I think leaving it as a simple `bool` for now is simpler.
</blockquote></details>

Cast isn't necessary:

```go
func (lm *labelMaker) MetricLabels(envelope *events.Envelope, originChoice OriginOption) map[string]string {
  if originChoice == WithOrigin {
    labels.setIfNotEmpty("origin", envelope.GetOrigin())
  }
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161)*
<!-- Sent from Reviewable.io -->

knyar commented:
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 81 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXmJJ_gr6Rx5gTLZZ:-L-Xfvz0AkfQFV3JBU-5:bgeywu9) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L81)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

Cast isn't necessary:

```go
func (lm *labelMaker) MetricLabels(envelope *events.Envelope, originChoice OriginOption) map[string]string {
  if originChoice == WithOrigin {
    labels.setIfNotEmpty("origin", envelope.GetOrigin())
  }
}
```
</blockquote></details>

Sorry, I meant either casting the boolean returned by `isRuntimeMetric()` into an `OriginOption`,  or branching out like this:
```
var labels map[string]string
if ms.isRuntimeMetric(envelope) {
  labels = ms.labelMaker.MetricLabels(envelope, nozzle.WithOrigin)
} else {
  labels = ms.labelMaker.MetricLabels(envelope, nozzle.WithoutOrigin)
}
```
(though retrospectively this actually looks quite good)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/stackdriver-nozzle/nozzle/label_maker.go, line 81 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161#-L-XXmJJ_gr6Rx5gTLZZ:-L-XiFQxifvvCDkb2B3H:b50yzj) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/07ad05108ccc43166491731f27723dc576495a64/src/stackdriver-nozzle/nozzle/label_maker.go#L81)):*
<details><summary><i>Previously, knyar (Anton Tolchanov) wrote…</i></summary><blockquote>

Sorry, I meant either casting the boolean returned by `isRuntimeMetric()` into an `OriginOption`,  or branching out like this:
```
var labels map[string]string
if ms.isRuntimeMetric(envelope) {
  labels = ms.labelMaker.MetricLabels(envelope, nozzle.WithOrigin)
} else {
  labels = ms.labelMaker.MetricLabels(envelope, nozzle.WithoutOrigin)
}
```
(though retrospectively this actually looks quite good)
</blockquote></details>

ah- my bad! A million ways to make the code shine :) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/161)*
<!-- Sent from Reviewable.io -->
